### ISSUE TITLE
Docs in sync-gateway is delayed to show up when DB brought back online from offline

### ISSUE THREAD

**Issue by sridevi-15** [couchbase/sync_gateway#2967]
### Sync Gateway version
1.5.0-594, 1.5.0-589 

### Couchbase server version
4.x -> tested with 4.6.3, 4.6.4-4544

### Operating system
CentOS

### Config file
```
{
    "out.stdout": {
        "adminInterface": "0.0.0.0:4985", 
        "compressResponses": false, 
        "databases": {
            "db": {
                "bucket": "data-bucket", 
                "offline": true, 
                "password": "password", 
                "server": "http://172.23.106.160,172.23.105.57,172.23.107.178:8091", 
                "username": "data-bucket", 
                "users": {
                    "seth": {
                        "admin_channels": [
                            "*", 
                            "ABC"
                        ], 
                        "password": "password"
                    }
                }
            }
        }, 
        "interface": ":4984", 
        "log": [
            "CRUD+", 
            "Cache+", 
            "HTTP+", 
            "Changes+"
        ], 
        "maxCouchbaseConnections": 16, 
        "maxFileDescriptors": 90000, 
        "maxIncomingConnections": 0, 
        "slowServerCallWarningThreshold": 500, 
        "verbose": "true"
    }
}

```

### Log output
2017-10-11T10:37:16.863-07:00 WARNING: Unexpected error parsing body of rev "1-c3bcf352dc26b1790d78cfe030738faf" -- db.(*document).getNonWinningRevisionBody() at document.go:309
2017-10-11T10:37:16.863-07:00 CRUD+: Got old revision "5d97b12e-8f12-4ede-9593-efdf048a6e80" / "1-c3bcf352dc26b1790d78cfe030738faf" --> 32 bytes
2017-10-11T10:37:16.863-07:00 Changes+: MultiChangesFeed sending {Seq:3, ID:5d97b12e-8f12-4ede-9593-efdf048a6e80, Changes:[map[rev:1-c3bcf352dc26b1790d78cfe030738faf]]}   (to seth)
2017-10-11T10:37:16.863-07:00 Changes+: Channel feed processing seq:7 in channel ABC   (to seth)
2017-10-11T10:37:16.864-07:00 WARNING: Unexpected error parsing body of rev "1-c3bcf352dc26b1790d78cfe030738faf" -- db.(*document).getNonWinningRevisionBody() at document.go:309
2017-10-11T10:37:16.864-07:00 CRUD+: Got old revision "ed3a77c2-8557-426a-8489-1b3439b70f22" / "1-c3bcf352dc26b1790d78cfe030738faf" --> 32 bytes
2017-10-11T10:37:16.864-07:00 Changes+: MultiChangesFeed sending {Seq:4, ID:ed3a77c2-8557-426a-8489-1b3439b70f22, Changes:[map[rev:1-c3bcf352dc26b1790d78cfe030738faf]]}   (to seth)
2017-10-11T10:37:16.864-07:00 Changes+: Channel feed processing seq:8 in channel ABC   (to seth)
2017-10-11T10:37:16.864-07:00 WARNING: Unexpected error parsing body of rev "1-c3bcf352dc26b1790d78cfe030738faf" -- db.(*document).getNonWinningRevisionBody() at document.go:309

### Steps to reproduce
1.  Set up CBS 4.6.4-4544, set sync-gateway 1.5.0-589 on Centos
2.  Have db offline on sync-gateway, wait for 10 seconds
3.  Bring db online , wait for 20 seconds
4. Hit some sync-gateway end point like
    Create role, create user, Create docs, update docs , create local docs, get local doc, 
5. Create 100 individual docs
6. Create 100 with add_bulk_doc api.
7. Create again 100 individual docs.
8. update all 300 docs
8. get bulk_docs, get _all_docs, 
9. Sleep for 2 seconds
10. Verify changes using _changes API  and verify number of docs, revisions prefix

 ### Expected behavior
Changes API should give 300 number of docs and with revision prefix 2-

### Actual behavior
Changes API giving just 1 doc

When I put sleep time to 5 to 6 seconds, test passes 
This is happening with couchbase server 4.x , seeing this issue on 4.6.3 and 4.6.4-4544
This test passes with couchbase server 5.x

Logs attached : 
Logs when test failed 
[sync_gateway_error.log](https://github.com/couchbase/sync_gateway/files/1379472/sync_gateway_error.log)

Logs when test passed : 
[sync_gateway_error.log](https://github.com/couchbase/sync_gateway/files/1379473/sync_gateway_error.log)

Failed Jenkins job
http://uberjenkins.sc.couchbase.com:8080/job/cen7-sync-gateway-functional-tests-base-cc/1177/testReport/

**Comment by adamcfraser** [comment]
It looks like the additional latency has to do with SG sequence buffering - in the failing case, SG is blocked waiting for sequence #51, and unblocks once that arrives.  

However, it looks like the 'missing' sequence is actually arriving much earlier and isn't being properly buffered:
https://gist.github.com/adamcfraser/e92df5b8e78595255e3b112257a1567d

Sequence 51 is originally being buffered because sequence 50 hasn't arrived, but would expect this to be cached when 50 is cached (at 11:16:27.076).

**Comment by adamcfraser** [comment]
Looking closer, the underlying delay seems to be caching of sequence 8 - the test waits around six seconds for that to show up:
```
2017-10-10T11:16:21.672-07:00 Cache:   Deferring #95 (110 now waiting for #8...#16)
...
2017-10-10T11:16:27.075-07:00 Cache:   Deferring #309 (380 now waiting for #8...#16)
```

**Comment by adamcfraser** [comment]
Sequence 8 is showing up as expected on the feed while the feed is still buffering seq 6, but I don't see the expected log message showing that seq 8 is deferred:
```
2017-10-10T11:16:21.204-07:00 Cache: Received #7 after   0ms ("30059b07-1261-405a-986a-7c19560a441b" / "1-c3bcf352dc26b1790d78cfe030738faf")
2017-10-10T11:16:21.204-07:00 Cache:   Deferring #7 (1 now waiting for #6...#6)
...
2017-10-10T11:16:21.227-07:00 Cache: Received #8 after  22ms ("cb6d1a18-ae50-42dc-a949-5e72ed9a0495" / "1-c3bcf352dc26b1790d78cfe030738faf")
...
2017-10-10T11:16:21.671-07:00 Cache:   Deferring #167 (110 now waiting for #6...#6)
2017-10-10T11:16:21.672-07:00 Cache: Received #6 after 468ms ("c3bb61d9-ec15-4596-a62b-1887f66d0cd4" / "1-c3bcf352dc26b1790d78cfe030738faf")
2017-10-10T11:16:21.672-07:00 Cache:     #6 ==> channel "ABC"
2017-10-10T11:16:21.672-07:00 Cache:     #6 ==> channel "*"
2017-10-10T11:16:21.672-07:00 Cache:     #7 ==> channel "ABC"
2017-10-10T11:16:21.672-07:00 Cache:     #7 ==> channel "*"
```

We expect to see a line like `Cache:   Deferring #8 (1 now waiting for #6...#6)` but it's not seen.

**Comment by adamcfraser** [comment]
This error occurs prior to the 'Deferring' message failing to show up for sequence 8:
```
2017-10-10T11:16:21.221-07:00 WARNING: changeCache: Error unmarshaling doc "_sync:role:radio_stations": Attempt to unmarshal principal using closed bucket -- db.(*changeCache).processPrincipalDoc() at change_cache.go:530
```
This is an unusual error - the bucket for the database context associated with the change cache is nil.  One possibility is that there are multiple change cache instances running (related to the offline/online changes), but we don't see duplicate Cache logging that we'd expect in that scenario.

**Comment by adamcfraser** [comment]
Some of this problem may be related to the first DCP stream (prior to offline/online) not being terminated.  There have been fixes related to this already applied to master (for 1.5.1).

@sridevi-15 Would you be able to rerun the test against the latest 1.5.1 build, and if it's still failing, attach updated logs to this ticket?

**Comment by sridevi-15** [comment]
Adam, tests has passed with 1.5.1-625 with CBS 4.6.3 and 4.6.4-4554

**Comment by adamcfraser** [comment]
Thanks Sridevi - based on that, I'm going to close this as fixed in 1.5.1 by the change listener termination fixes included with https://github.com/couchbase/sync_gateway/commit/6447c918d7573fc0eb362b4b5d3cec9b62c9e0e4.

**Comment by nbrys** [comment]
We saw the same error messages yesterday on our cluster with sync gateway 1.5.1. The difference is that our CB cluster didn't went offline.

> Feb 01 16:44:27 mobile-prod-syncgateway-i-050fc70db358efd5a sync-gateway-stdout: 2018-02-01T15:44:27.034Z HTTP:  #2157084: GET /customerdb/d255755c-67f7-432d-9250-29c2c4dbfdd9/default.xml  (as documentflow)
Feb 01 16:44:27 mobile-prod-syncgateway-i-050fc70db358efd5a sync-gateway-stderr: 2018-02-01T15:44:27.065Z HTTP:  #2157085: POST /customerdb/_changes?feed=longpoll&heartbeat=30000&style=all_docs&since=2941479&filter=sync_gateway%2Fbychannel  (as 10592)
Feb 01 16:44:27 mobile-prod-syncgateway-i-0efd9afeaf26f9e5f sync-gateway-stderr: _time=2018-02-01T15:44:27.502+00:00 _level=INFO _msg=go-couchbase: call to ViewCustom("sync_gateway", "channels") in github.com/couchbase/sync_gateway/db.(*DatabaseContext).getChangesInChannelFromView took 2.850778897s 
Feb 01 16:44:27 mobile-prod-syncgateway-i-0efd9afeaf26f9e5f sync-gateway-stdout: 2018-02-01T15:44:27.569Z changes_view: Query took 2.918115721s to return 125996 rows, options = db.Body{"stale":false, "startkey":[]interface {}{"__status_outbox", 0x2}, "endkey":[]interface {}{"__status_outbox", 0x2ce175}}
Feb 01 16:44:27 mobile-prod-syncgateway-i-0efd9afeaf26f9e5f sync-gateway-stderr: 2018-02-01T15:44:27.578Z WARNING: Unexpected error parsing body of rev "5-e49c54121b945d0aa2f9bc491307a10b" -- db.(*document).getNonWinningRevisionBody() at document.go:309
Feb 01 16:44:27 mobile-prod-syncgateway-i-0efd9afeaf26f9e5f sync-gateway-stdout: 2018-02-01T15:44:27.579Z WARNING: Changes feed: error getting doc "-WTRPIu_iqzx2tRdhBFvTN6"/"5-e49c54121b945d0aa2f9bc491307a10b": 404 missing -- db.(*Database).AddDocInstanceToChangeEntry() at changes.go:141
Feb 01 16:44:27 mobile-prod-syncgateway-i-0efd9afeaf26f9e5f sync-gateway-stderr: 2018-02-01T15:44:27.580Z WARNING: Unexpected error parsing body of rev "6-26974a452e33789c95891a490d9765c9" -- db.(*document).getNonWinningRevisionBody() at document.go:309

**Comment by tleyden** [comment]
@nbrys can you open a new issue and fill out the template?  Also, post any short logs within "triple backticks" (see https://guides.github.com/features/mastering-markdown/) .. and anything over 50 lines or so in a gist.
